File: courserapy/dataimport.py
from os import getcwd
CWD = getcwd()
DATDIR = '/datfiles/'


class masstable(object):
    '''Canned lists and methods for amino acid masses in peptide sequencing problems'''

    masses = list()
    acids = list()
    __mass_dict = dict()

    def __init__(self):
        self.parse(CWD + DATDIR + 'masstable.txt')

    def parse(self, file):

        with open(file, 'r') as f:
            data = [line.strip().split(' ') for line in f.readlines()]
        for item in data:
            self.acids.append(item[0])
            self.masses.append(int(item[1]))
            self.__mass_dict[item[0]] = int(item[1])

    def get_mass(self, acid):
        return self.__mass_dict[acid]


class Score(object):
    '''Canned scoring matrices for amino acid sequence alignment problems'''
    NEGATIVE_IDENTITY = dict()
    INVERSE_IDENTITY = dict()
    IDENTITY = dict()
    ALL_ONES = dict()
    BLOSUM62 = dict()
    PAM250 = dict()

    def __init__(self):
        # identity matrices for all amino acids
        self.NEGATIVE_IDENTITY = Score.dictify(
            CWD + DATDIR + 'NEGATIVE_IDENTITY.txt')
        self.INVERSE_IDENTITY = Score.dictify(
            CWD + DATDIR + 'INVERSE_IDENTITY.txt')
        self.IDENTITY = Score.dictify(CWD + DATDIR + 'IDENTITY.txt')
        self.ALL_ONES = Score.dictify(CWD + DATDIR + 'ALL_ONES.txt')
        # biologically significant matrices
        self.BLOSUM62 = Score.dictify(CWD + DATDIR + 'BLOSUM62.txt')
        self.PAM250 = Score.dictify(CWD + DATDIR + 'PAM250.txt')

    @staticmethod
    def isnumeric(s):
        ''' builtin str.isnumeric() does not account for '-' '''
        l = "-0123456789"
        return all([x in list(l) for x in list(s)])

    @staticmethod
    def dictify(file):
        '''Canned method to create dict of dicts...definitely better ways to do this'''
        with open(file, 'r') as input_data:
            data = [[int(x) if Score.isnumeric(x) else x
                     for x in line.strip().split(' ') if x != '']
                    for line in input_data.readlines()]
        data_dict = dict()
        matrix = [[[let, [data[j][0], data[i + 1][j]]]
                   for j in range(1, len(data))]
                  for i, let in enumerate(data[0])]
        for item in matrix:
            for k, kv in item:
                if k in data_dict:
                    l = [[key, data_dict[k][key]] for key in data_dict[k]]
                    l.append([kv[0], kv[1]])
                    d = {}
                    d[k] = {ky: vl for ky, vl in l}
                    data_dict.update(d)
                else:
                    data_dict[k] = {kv[0]: kv[1]}
        return data_dict

if __name__ == "__main__":
    test = Score().INVERSE_IDENTITY
    print(test)

File: courseracsharp/DataImport.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

namespace Coursera
{
    public class MassData
    {

        static Dictionary<char, int> acid_to_mass = new Dictionary<char, int>();
        static Dictionary<int, char> mass_to_acid = new Dictionary<int, char>();
        public MassData()
        {
            //string path = Directory.GetCurrentDirectory() + "\\masstable.txt";
            string cwd = Directory.GetCurrentDirectory();
            string path = cwd + "//masstable.txt";
            FileStream stream = new FileStream(path, FileMode.Open);
            StreamReader file = new StreamReader(stream);
            parse(file);
        }
        static void parse(StreamReader file)
        {
            foreach (string line in file.ReadToEnd().Split('\n'))
            {
                string[] temp = line.Split(' ');
                acid_to_mass[temp[0][0]] = int.Parse(temp[1]);
            }
            foreach (KeyValuePair<char, int> x in acid_to_mass)
            {
                if (!mass_to_acid.ContainsKey(x.Value)) mass_to_acid[x.Value] = x.Key;
            }
        }
        public int get_mass(char acid) { return acid_to_mass[acid]; }
        public char get_acid(int mass) { return mass_to_acid[mass]; }
        public List<int> masses() { return mass_to_acid.Keys.ToList(); }
        public List<char> acids() { return acid_to_mass.Keys.ToList(); }
    }
    public class Score
    {
        public Dictionary<char, Dictionary<char, int>> NEGATIVE_IDENTITY;
        public Dictionary<char, Dictionary<char, int>> INVERSE_IDENTITY;
        public Dictionary<char, Dictionary<char, int>> IDENTITY;
        public Dictionary<char, Dictionary<char, int>> ALL_ONES;
        public Dictionary<char, Dictionary<char, int>> BLOSUM62;
        public Dictionary<char, Dictionary<char, int>> PAM250;
        public Score()
        {
            string cwd = Directory.GetCurrentDirectory();
            string datdir = "\\datfiles\\";
            string path = cwd + datdir + "//masstable.txt";
            FileStream stream = new FileStream(path, FileMode.Open);
            StreamReader file = new StreamReader(stream);

            // pass streamreader to dictify function and return Dict<char,char> using txt file names as arguments
            Func<string, Dictionary<char, Dictionary<char, int>>> get_dictionary = delegate (string f)
            { return this.dictify(new StreamReader(new FileStream(cwd + datdir + f, FileMode.Open))); };

            this.NEGATIVE_IDENTITY = get_dictionary("NEGATIVE_IDENTITY.txt");
            this.INVERSE_IDENTITY = get_dictionary("INVERSE_IDENTITY.txt");
            this.IDENTITY = get_dictionary("IDENTITY.txt");
            this.ALL_ONES = get_dictionary("ALL_ONES.txt");
            // biologically significant matrices
            this.BLOSUM62 = get_dictionary("BLOSUM62.txt");
            this.PAM250 = get_dictionary("PAM250.txt");
        }
        Dictionary<char, Dictionary<char, int>> dictify(StreamReader datafile)
        {
            Dictionary<char, Dictionary<char, int>> dict = new Dictionary<char, Dictionary<char, int>>();
            string[][] data = datafile.ReadToEnd().Split(new char[] { '\n' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(x => x.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries)).ToArray();
            for (int i = 0; i < data[0].Count(); i++)
            {
                for (int j = 1; j < data.Count(); j++)
                {
                    Dictionary<char, int> temp;
                    if (dict.ContainsKey(data[0][i][0])) temp = dict[data[0][i][0]];
                    else temp = new Dictionary<char, int>();

                    temp.Add(data[j][0][0], int.Parse(data[i + 1][j]));
                    dict[data[0][i][0]] = temp;
                }
            }
            return dict;
        }
    }
}